### ISSUE TITLE
feat: Python Graph Dataset API for Data Science consumption (#773)

### ISSUE THREAD

**Issue by danielperezr88** [danielperezr88/astrolabe#804]
## Summary

Create a Python package (`astrolabe-datasets`) that exposes Astrolabe knowledge graph data as rich graph datasets consumable by major Data Science and ML frameworks. This enables researchers and practitioners to train GNNs, run graph analytics, and perform statistical analysis on real-world code knowledge graphs.

## Motivation

Astrolabe generates rich heterogeneous code knowledge graphs (49 node types, 26 edge types, computed metrics, community structure). Currently this data is locked in SQLite and accessible only via the Node.js API/MCP tools. A Python API unlocks:

- GNN training for code smell detection, vulnerability detection, maintainability prediction
- Statistical analysis of code graph topology across repositories
- Benchmarking against CodeXGLUE, Devign, BigCloneBench datasets
- Graph visualization and exploratory analysis in Jupyter notebooks

### 1. PyTorch-Geometric (PyG) — `CodeGraphDataset`

- `InMemoryDataset` subclass with `HeteroData` containers
- Lazy and eager loading modes
- Node features per type: type encoding, degree, visibility, parameter count, async/static flags
- Edge features per type: confidence, step ordinal
- Computed metrics as additional node features: PageRank, betweenness centrality, community cohesion
- Heterogeneous schema: `(src_node_type, edge_type, dst_node_type)` triplets
- Normalization of missing types across graphs (critical PyG requirement)
- Support for `HGTConv`, `HeteroConv`, and `to_hetero()` model patterns
- Export to OGB format for benchmark interop

**Key implementation note**: Every `HeteroData` in the dataset must have identical node/edge type schemas. Empty tensors must be added for missing types.

### 2. NVIDIA RAPIDS cuGraph — `CugraphBackend`

- GPU-accelerated graph construction from cuDF DataFrames
- Support for `cugraph.hypergraph()` with NODETYPE/EDGETYPE columns
- Expose all cuGraph algorithms: Leiden, Louvain, Katz centrality, SSSP, HITS, Jaccard link prediction, Node2Vec
- Integration with `cugraph_pyg` for GPU-native PyG GraphStore/FeatureStore/Sampler
- Multi-GPU support via Dask-cuDF for large-scale analysis

**cuGraph status**: Production-grade (v26.04), Apache 2.0, NVIDIA-backed. Heterogeneous graphs supported via property graph API. PyG-only (DGL support removed as of 25.06).

### 3. Apache Spark GraphFrames — `SparkGraphFrameBackend`

- PropertyGraphFrame API (v0.11.0, Apr 2026) for heterogeneous vertex/edge types
- VertexPropertyGroup per node type, EdgePropertyGroup per edge type
- Distributed algorithms: PageRank, connected components, label propagation, BFS, shortest paths, triangle count, K-core
- RandomWalkEmbeddings (Word2Vec/Hash2Vec) for graph embeddings
- PySpark DataFrame-native representation for ETL pipelines
- Spark Connect compatible (Classic + Connect protocols)

### 4. Pandas Ecosystem — Multi-backend

- **igraph** (`Graph.DataFrame`): C-core, fast, `get_edge_dataframe()`/`get_vertex_dataframe()` round-trip
- **graph-tool**: OpenMP parallelism, Stochastic Block Model community detection, fastest single-machine option, filtered graph views
- **NetworkX**: `from_pandas_edgelist()`/`to_pandas_edgelist()` for prototyping

### 5. Unified Export Formats

- OGB dict format (framework-agnostic, works with PyG and DGL)
- GraphML/GML (portable across all graph libraries)
- Edge/Node CSV (universal baseline)

## Data Model

All backends consume from the same canonical representation:

```python
@dataclass
class CodeGraph:
    nodes: pd.DataFrame       # columns: node_id, label, name, filePath, startLine, endLine, language, + all NodeProperties
    edges: pd.DataFrame       # columns: edge_id, source_id, target_id, type, confidence, reason, step
    embeddings: pd.DataFrame  # columns: node_id, vector (numpy array), dims
    metrics: pd.DataFrame     # columns: node_id, pagerank, betweenness, community_id, cohesion
    health: dict              # ArchitectureHealth: overallScore, cohesion, modularity, complexity, antiPatterns
    graphlets: dict           # GraphletProfile: motif counts
```

## SQLite Ingestion

The API reads directly from Astrolabe SQLite files:
- `nodes` table: id, label, properties (JSON)
- `relationships` table: id, source_id, target_id, type, confidence, reason, step, evidence (JSON)
- `embeddings` table: node_id, hash, vector (BLOB/Float32), dims, indexed_at
- `fts_nodes` virtual table: node_id, label, name, filePath, keywords

## Package Structure

```
astrolabe-datasets/
├── pyproject.toml
├── astrolabe_datasets/
│   ├── __init__.py
│   ├── core.py              # CodeGraph dataclass, SQLite loader
│   ├── lazy.py              # Lazy loaders (iterator-based)
│   ├── pyg_dataset.py       # PyG InMemoryDataset + HeteroData
│   ├── cugraph_backend.py   # cuGraph GPU backend
│   ├── spark_backend.py     # GraphFrames distributed backend
│   ├── pandas_backends.py   # igraph, graph-tool, NetworkX
│   ├── exporters.py         # OGB, GraphML, CSV exporters
│   └── feature_engineering.py  # Node/edge feature encoding utilities
└── tests/
```

## Labels

- **Category**: `python-api`, `graph-analytics`
- **Effort**: `large` (1-2 months)
- **Risk**: `medium-risk` (well-understood library APIs, complexity in multi-backend normalization)
- **Benefit**: Enables Astrolabe as a platform for academic research and production graph ML

## Dependencies

- `astrolabe-core` (SQLite file format compatibility)
- `torch`, `torch-geometric` (PyG backend)
- `cudf`, `cugraph`, `cugraph-pyg` (cuGraph backend, optional)
- `pyspark`, `graphframes` (Spark backend, optional)
- `igraph`, `graph-tool`, `networkx` (Pandas backends)
- `pandas`, `numpy` (core)

**Comment by danielperezr88** [comment]
Closed by PR #814 (merged to staging).
